### PR TITLE
runs on docker

### PR DESCRIPTION
**Pull request by Ognian**
- added dockerfile and docker-compose
- moved from kpartx to partprobe for docker compatibility

### CODE REVIEW

**Comment by sgielen** [comment]
Hi @Ognian, sorry for not getting back to you earlier. I'm reviewing the script today.

I'll test if your changes don't break anything in the normal build so far, and will also double-check whether I can also build the image in Docker for Mac now.

I have some small comments up-front, could you perhaps address those? I promise I'll be quicker in the second round :)

File: README.md [hunk 1]
```diff
@@ -49,6 +49,11 @@ the normal init script and try to start k3os manually. You may need to load addi
 
 Anytime you think the scripts or documentation could be improved, please file an issue or a PR and we'll be happy to help.
 
+## Docker
+If you want to build in Docker for Mac you can use the `docker-compose.yml` in the `docker-k3os-image-gen directory`.
+The `images` subdirectory will contain the resulting image. The `<imagetype>` is set via the `TARGET` envirnonment in the `docker-compose.yml` file.
```
Comment (sgielen) [inline README.md:54]: (Typo, environment)

File: build-image.sh [hunk 2]
```diff
@@ -148,15 +148,22 @@ elif [ "$IMAGE_TYPE" = "orangepipc2" ]; then
 fi
 
 LODEV=`sudo losetup --show -f $IMAGE`
-sudo kpartx -a $LODEV
+echo $LODEV
```
Comment (sgielen) [inline build-image.sh:151]: Please remove debugging output (in multiple places)

File: build-image.sh [hunk 3]
```diff
@@ -148,15 +148,22 @@ elif [ "$IMAGE_TYPE" = "orangepipc2" ]; then
...
+
+#sudo kpartx -a $LODEV
```
Comment (sgielen) [inline build-image.sh:153]: Please remove commented out code

File: build-image.sh [hunk 4]
```diff
@@ -148,15 +148,22 @@ elif [ "$IMAGE_TYPE" = "orangepipc2" ]; then
...
+cat /proc/partitions | grep loop
+ls -al /dev/loop*
+
 if [ "$IMAGE_TYPE" = "raspberrypi" ]; then
-	LODEV_BOOT=/dev/mapper/`basename ${LODEV}`p1
-	LODEV_ROOT=/dev/mapper/`basename ${LODEV}`p2
-	sudo mkfs.fat $LODEV_BOOT
+	LODEV_BOOT=${LODEV}p1
+	LODEV_ROOT=${LODEV}p2
+  sudo mkfs.fat $LODEV_BOOT
```
Comment (sgielen) [inline build-image.sh:164]: Please rollback unnecessary changes because of indentation differences

File: burn-image-osx.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (sgielen) [inline burn-image-osx.sh:1]: I would prefer to have a `burn-image.sh` that checks whether the platform is OS X, so we can add support for different platforms later (i.e. provide an alternative to the `diskutil` commands)

File: docker-k3os-image-gen/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    environment:
+      TARGET: raspberrypi
+    volumes:
+      # bind ../config to our working dir config
+      # this ONLY works with unnamed volumes like below!!
+      - ../config:/usr/src/app/config
+      # result is in this named volume
+      - "images:/usr/src/app/images"
+      # this is the downloads cache
+      - "k3os-fw:/usr/src/app/deps"
```
Comment (sgielen) [inline docker-k3os-image-gen/docker-compose.yml:14]: Having the downloads cache in a named volume is fine by me, but why is there a difference in how you bind the `config` directory and how you bind the `images` directory?

Comment (sgielen) [reply docker-k3os-image-gen/docker-compose.yml:14]: Wouldn't it be easy & acceptable if we just bind the entire repository into the Docker container?

File: docker-k3os-image-gen/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    devices:
+      - "/dev:/dev"
+    cap_add:
+      - CAP_MKNOD
+volumes:
+   k3os-fw:
+     driver: local
+   images:
+     # OSX only!! if we add this the below path under device is where the volume will reside actually on OSX
+     # otherwise it is in the docker for OSX vm, where we have ONLY indirect access via other containers or command line ...
```
Comment (sgielen) [inline docker-k3os-image-gen/docker-compose.yml:27]: This would, in principle, be supported for any OS, right? Not just Docker for Mac? E.g. Docker for Windows, and just native Linux Docker, would also allow building this way? In this case, perhaps we could remove the references to OSX and just mention in the README that this is a possible way to build the image on Windows, Mac or Linux without needing local dependencies?

**Comment by sgielen** [comment]
By the way, you moved from `kpartx` to `partprobe` so that the partitions are at `/dev/loopNpM` instead of `/dev/mapper/...`, but with all of `/dev` bind-mounted, do you still need that?

File: docker-k3os-image-gen/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+version: "3"
```
Comment (sgielen) [inline docker-k3os-image-gen/docker-compose.yml:1]: Was there any particular reason you put the docker-compose file in a separate directory, but the Dockerfile in the root?

File: docker-k3os-image-gen/docker-compose.yml (hunk 6 above)
Comment (Ognian) [reply docker-k3os-image-gen/docker-compose.yml:14]: the `config` directory is an input for the docker container; the way I do it here is the only way I know how the config directory is read every time, even if the images for the container already exist. The `images` and `cache` directory are output directories and need to persist, therefore a chooser named volumes and since under OSX and Windows they are actually in a VM we need to bind mount them ... Which also answers the question why docker-compose is in a different directory: since depending on which platform it runs (OSX, Windows or Linux) we may need different docker-compose ...

File: docker-k3os-image-gen/docker-compose.yml (hunk 7 above)
Comment (Ognian) [reply docker-k3os-image-gen/docker-compose.yml:27]: named volumes with bind mount work perfectly under OSX, somehow under Windows and are not needed for Linux ... :-)
Maybe this changed in the last months, since there is now a new option [mutagen chaching](https://docs.docker.com/docker-for-mac/mutagen-caching/ ) but testing this on all 3 platforms is a lot of work...

File: docker-k3os-image-gen/docker-compose.yml (hunk 8 above)
Comment (Ognian) [reply docker-k3os-image-gen/docker-compose.yml:1]: we may need different docker-compose for the different platforms...
the dockerfile will stay the same

**Comment by Ognian** [comment]
> By the way, you moved from `kpartx` to `partprobe` so that the partitions are at `/dev/loopNpM` instead of `/dev/mapper/...`, but with all of `/dev` bind-mounted, do you still need that?
> 
> **Edit:** It doesn't seem to make any difference in my existing builds, and `partprobe` is installed together with `parted` while `kpartx` is separate so it saves a dependency. I also like how we don't need that `/dev/mapper/$(basename...)` stuff anymore. If you're OK with it, I might commit that part separately to master, with credits to you in the commit message of course, and then I believe you wouldn't actually need to make any changes to the `build-image.sh` itself in that case. :)

The reason I moved from `kpartx` to` partprobe` was that it simply didn't work with the docker desktop version at that time. With the newest one this should be resolved...
Fell free to make the changes how it suits you best

**Comment by Ognian** [comment]
@sgielen I saw you already merged the partprobe but not the docker stuff... When are you going to do that?

**Comment by sgielen** [comment]
> @sgielen I saw you already merged the partprobe but not the docker stuff... When are you going to do that?

I was porting your changes to a minimal commit, including only the Dockerfile and README changes, as there are quite some changes in this PR that I think aren't necessary. For example, adding only whitespaces or creating an `.editorconfig`.

I have the build running on Linux, but was testing in Docker for Mac, where this issue occurs:

```
== Making image and filesystems... ==
fallocate: fallocate failed: Operation not supported
```

It looks like [the filesystem used by Docker doesn't support fallocate](https://stackoverflow.com/questions/31155591/fallocate-failing-inside-my-docker-container). Probably, your way of mounting the volume prevents this issue from occurring, but I am hoping for a solution that does not depend on the host filesystem...

**Comment by sgielen** [comment]
Actually, I've just tried with truncate, and that works a lot better. I'll create a separate commit for that and add some instructions for building on the Mac to the README.

**Comment by sgielen** [comment]
- Switch from kpartx to partprobe: 9e3f1b7c70e5275b5e03abf6c866a88c4d0bca28
- Allow building in Docker: abd9ca3ec5f6bd2d007d3685f72396d86f8f150a
- Switch from fallocate to truncate: 6a0f20ab1847c33bb45a654de1b45cfa3b42de3e

With these three, mostly contributed by you (with credit & many thanks!), building now works on master with Docker for Mac.

I've made some changes to how the building works; it mainly doesn't need docker-compose. From the README:

> You can build this project in Docker, e.g. when you'd rather not install dependencies on your host machine or when you're
> building on a Mac.
> 
> If you want to build in Docker, you can build a container containing the dependencies using:
> 
> ```
> docker build . -t picl-builder:latest
> ```
> 
> Then, run the container using:
> 
> ```
> docker run -e TARGET=all -v $(realpath .):/app -v /dev:/dev --privileged picl-builder:latest
> ```
> 
> The images will be written into your local directory once the container is done.

**Comment by sgielen** [comment]
From this PR, that leaves your Docker-compose.yml and `burn-image.sh`.

Do you think there is still a good reason to include the docker-compose file on top of what is already on master now?

For the burn-image.sh, I'd like to have that in master as well. The images are written just to the repository root on master, not to the `images/` directory, so it needs tiny changes, and I'd still like to merge it without the other unrelated changes mentioned before. How do you think we should best do that?

**Comment by Ognian** [comment]
@sgielen I added a new pull request with the modified burn-image.sh so you can merge it quickly.
For now the description and the docker file should be enough, if I find a good reason for the docker-compose file we can add it later on...
So we can close this pull request.
